os/wqueue/kwqueue/kwork_signal.c

```diff
@@ -88,5 +91,21 @@ int work_signal(int qid)
 		return -EINVAL;
 	}
 
-	return work_qsignal(pid);
+	/* Signal must be sent only when worker thread is sigwait state.
+	 * Otherwise, unnecessary signal will be sent(when worker thread is
+	 * processing the work queue).
+	 */
+
+	state = task_getstate(pid);
+	if (state == TSTATE_WAIT_SIG) {
```

Comment:
I am not sure if this check is required.
We already have a check to prevent nesting of signals in up_schedulesigaction.c file as below.

```
        /* Refuse to handle nested signal actions */

        if (!tcb->xcp.sigdeliver) {
                ...
        }
```